infer/openapipulschema/provider.go

```diff
@@ -0,0 +1,94 @@
+// Copyright 2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package openapipulschema
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+
+	pulschema "github.com/cloudy-sky-software/pulschema/pkg"
+	"github.com/getkin/kin-openapi/openapi3"
+	p "github.com/pulumi/pulumi-go-provider"
+	"github.com/pulumi/pulumi-go-provider/middleware/schema"
+	pschema "github.com/pulumi/pulumi/pkg/v3/codegen/schema"
+)
+
+const packageName = "openapi"
+
+func Wrap(provider p.Provider, oaURL url.URL, metadata schema.Metadata) (p.Provider, error) {
```

Comment:
This can't actually `Wrap` a provider since it fully overrides the `GetSchema` method (instead of merging with downstream schemas).

```suggestion
func Provider(oaURL url.URL, metadata schema.Metadata) (p.Provider, error) {
```

If you want to implement a `Wrap` function, I believe you will need to implement:

```go
// package middleware/schema
func FromExistingSchema("github.com/pulumi/pulumi/pkg/v3/codegen/schema".PackageSpec) Options { ... }
```

You would then replace `provider.GetSchema = ...` with `provider = schema.Wrap(provider, schema.FromExistingSchema(pkg))`. That way any schema provided by `provider` would still show through unless it was overridden by the provider schema created from the openAPI spec.